Binding source generator emits bad code for tuples

For binding `List<(string Error, string MemberName)>`, the binding source generator produces code like this:

```csharp

        public static void BindCore(IConfiguration configuration, ref global::System.Collections.Generic.List<(string Error, string MemberName)> instance, bool defaultValueIfNotFound, BinderOptions? binderOptions)
        {
            foreach (IConfigurationSection section in configuration.GetChildren())
            {
                if (!string.IsNullOrEmpty(section.Value) && !section.GetChildren().Any())
                {
                    continue;
                }
                var value = new (string Error, string MemberName)(); // error CS8181
                instance.Add(value);
            }
        }
```

which produces an error:

`error CS8181: 'new' cannot be used with tuple type. Use a tuple literal expression instead.`

--- a/src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Suppressor.cs
+++ b/src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Suppressor.cs
@@ -12,7 +12,7 @@ namespace Microsoft.Extensions.Configuration.Binder.SourceGeneration
     public sealed partial class ConfigurationBindingGenerator
     {
         /// <summary>
-        /// Supresses false-positive diagnostics emitted by the linker
+        /// Suppresses false-positive diagnostics emitted by the linker
         /// when analyzing binding invocations that we have intercepted.
         /// Workaround for https://github.com/dotnet/roslyn/issues/68669.
         /// </summary>

--- a/src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/CoreBindingHelpers.cs
+++ b/src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/CoreBindingHelpers.cs
@@ -1180,7 +1180,10 @@ namespace Microsoft.Extensions.Configuration.Binder.SourceGeneration
 
                     if (strategy is ObjectInstantiationStrategy.ParameterlessConstructor)
                     {
-                        initExpr = $"new {typeFQN}()";
+                        // value tuple types will be declared with syntax like:
+                        //     (int, int) value = default;
+                        // This is to avoid using invalid syntax calling the parameterless constructor
+                        initExpr = type.IsValueTuple ? "default" : $"new {typeFQN}()";
                     }
                     else
                     {
@@ -1195,7 +1198,11 @@ namespace Microsoft.Extensions.Configuration.Binder.SourceGeneration
                     case InitializationKind.Declaration:
                         {
                             Debug.Assert(!memberAccessExpr.Contains('.'));
-                            _writer.WriteLine($"var {memberAccessExpr} = {initExpr};");
+                            // value tuple will be declared with syntax like:
+                            //     (int, int) value = default;
+                            // We need to specify the typeFQN as we assign the variable to default value.
+                            string declarationType = type.IsValueTuple ? typeFQN : $"var";
+                            _writer.WriteLine($"{declarationType} {memberAccessExpr} = {initExpr};");
                         }
                         break;
                     case InitializationKind.AssignmentWithNullCheck:

--- a/src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Specs/Types/TypeSpec.cs
+++ b/src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Specs/Types/TypeSpec.cs
@@ -17,6 +17,7 @@ namespace Microsoft.Extensions.Configuration.Binder.SourceGeneration
             (DisplayString, FullName) = type.GetTypeNames();
             IdentifierCompatibleSubstring = type.ToIdentifierCompatibleSubstring();
             IsValueType = type.IsValueType;
+            IsValueTuple = type is INamedTypeSymbol namedTypeSymbol && namedTypeSymbol.IsTupleType;
         }
 
         public TypeRef TypeRef { get; }
@@ -36,6 +37,8 @@ namespace Microsoft.Extensions.Configuration.Binder.SourceGeneration
         public string IdentifierCompatibleSubstring { get; }
 
         public bool IsValueType { get; }
+
+        public bool IsValueTuple { get; }
     }
 
     public abstract record ComplexTypeSpec : TypeSpec

--- a/src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/GeneratorTests.Helpers.cs
+++ b/src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/GeneratorTests.Helpers.cs
@@ -202,5 +202,18 @@ namespace Microsoft.Extensions.SourceGeneration.Configuration.Binder.Tests
             }
             return assemblies;
         }
+
+        public static byte[] CreateAssemblyImage(Compilation compilation)
+        {
+            MemoryStream ms = new MemoryStream();
+            var emitResult = compilation.Emit(ms);
+            if (!emitResult.Success)
+            {
+                // Explicit failures to include in the test output.
+                string errorMessage = string.Join(Environment.NewLine, emitResult.Diagnostics.Select(d => d.ToString()));
+                throw new InvalidOperationException(errorMessage);
+            }
+            return ms.ToArray();
+        }
     }
 }

--- a/src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/GeneratorTests.cs
+++ b/src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/GeneratorTests.cs
@@ -296,10 +296,10 @@ namespace Microsoft.Extensions.SourceGeneration.Configuration.Binder.Tests
         }
 
         /// <summary>
-            /// We binding the type "SslClientAuthenticationOptions" which has a property "CipherSuitesPolicy" of type "CipherSuitesPolicy". We can't bind this type.
-            /// This test is to ensure not including the property "CipherSuitesPolicy" in the generated code caused a build break.
-            /// </summary>
-            [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsNetCore))]
+        /// We binding the type "SslClientAuthenticationOptions" which has a property "CipherSuitesPolicy" of type "CipherSuitesPolicy". We can't bind this type.
+        /// This test is to ensure not including the property "CipherSuitesPolicy" in the generated code caused a build break.
+        /// </summary>
+        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsNetCore))]
         public async Task IgnoredUnBindablePropertiesTest()
         {
             string source = """

--- a/src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/GeneratorTests.cs
+++ b/src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/GeneratorTests.cs
@@ -281,7 +281,6 @@ namespace Microsoft.Extensions.SourceGeneration.Configuration.Binder.Tests
                 public class Settings
                 {
                     public List<(string Item1, string? Item2)>? Items { get; set; }
-                    // public List<string>? Items { get; set; }
                 }
                 """;
 

--- a/src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/GeneratorTests.cs
+++ b/src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/GeneratorTests.cs
@@ -257,11 +257,49 @@ namespace Microsoft.Extensions.SourceGeneration.Configuration.Binder.Tests
             }
         }
 
-        /// <summary>
-        /// We binding the type "SslClientAuthenticationOptions" which has a property "CipherSuitesPolicy" of type "CipherSuitesPolicy". We can't bind this type.
-        /// This test is to ensure not including the property "CipherSuitesPolicy" in the generated code caused a build break.
-        /// </summary>
         [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsNetCore))]
+        public async Task ListOfTupleTest()
+        {
+            string source = """
+                using Microsoft.Extensions.Configuration;
+                using System;
+                using System.Collections.Generic;
+
+                public class Program
+                {
+                    public static void Main()
+                    {
+                        ConfigurationBuilder configurationBuilder = new();
+                        IConfiguration config = configurationBuilder.Build();
+
+                        var settingsSection = config.GetSection("Settings");
+
+                        Settings options = settingsSection.Get<Settings>()!;
+                    }
+                }
+
+                public class Settings
+                {
+                    public List<(string Item1, string? Item2)>? Items { get; set; }
+                    // public List<string>? Items { get; set; }
+                }
+                """;
+
+            ConfigBindingGenRunResult result = await RunGeneratorAndUpdateCompilation(source, assemblyReferences: GetAssemblyRefsWithAdditional(typeof(ConfigurationBuilder), typeof(List<>)));
+            Assert.NotNull(result.GeneratedSource);
+            Assert.Empty(result.Diagnostics);
+
+            // Ensure the generated code can be compiled.
+            // If there is any compilation error, exception will be thrown with the list of the errors in the exception message.
+            byte[] emittedAssemblyImage = CreateAssemblyImage(result.OutputCompilation);
+            Assert.NotNull(emittedAssemblyImage);
+        }
+
+        /// <summary>
+            /// We binding the type "SslClientAuthenticationOptions" which has a property "CipherSuitesPolicy" of type "CipherSuitesPolicy". We can't bind this type.
+            /// This test is to ensure not including the property "CipherSuitesPolicy" in the generated code caused a build break.
+            /// </summary>
+            [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsNetCore))]
         public async Task IgnoredUnBindablePropertiesTest()
         {
             string source = """

